Asgard/Communications/Classes/CbusEventManager.cs

```diff
@@ -0,0 +1,246 @@
+﻿using System;
+using System.Collections.Concurrent;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Asgard.Data;
+using Asgard.Data.Interfaces;
+using Microsoft.Extensions.Logging;
+
+namespace Asgard.Communications
+{
+    public class CbusEventManager :
+        ICbusEventManager,
+        IDisposable
+    {
+        #region Fields
+
+        private readonly ICbusMessenger cbusMessenger;
+
+        private readonly ILogger logger;
+
+        private readonly ConcurrentDictionary<CbusEventKey, bool?> cbusEventStates = new();
+        private readonly ConcurrentDictionary<CbusEventKey, CbusEventCallback> cbusEventCallbacks = new();
+        private readonly EmptyingConcurrentQueue<Exception> cbusEventExceptionQueue = new();
+
+        private readonly CancellationTokenSource cancellationTokenSource = new();
+
+        #endregion
+
+        #region Properties
+
+        protected CancellationToken CancellationToken => this.cancellationTokenSource.Token;
+
+        #endregion
+
+        #region Constructors
+
+        public CbusEventManager(ICbusMessenger cbusMessenger, ILogger logger)
+        {
+            this.cbusMessenger = cbusMessenger;
+            this.logger = logger;
+
+            Task.Run(() => ProcessQueue());
+
+            this.cbusMessenger.MessageReceived += CbusMessenger_MessageReceived;
+        }
+
+        #endregion
+
+        #region IDisposable support
+
+        public bool IsDisposed { get; private set; }
+
+        protected virtual void Dispose(bool disposing)
+        {
+            if (!this.IsDisposed)
+            {
+                if (disposing)
+                {
+                    this.cancellationTokenSource.Cancel();
+                    // TODO: dispose managed state (managed objects)
+                }
+
+                // TODO: free unmanaged resources (unmanaged objects) and override finalizer
+                // TODO: set large fields to null
+                this.IsDisposed = true;
+            }
+        }
+
+        // // TODO: override finalizer only if 'Dispose(bool disposing)' has code to free unmanaged resources
+        // ~CbusEventManager()
+        // {
+        //     // Do not change this code. Put cleanup code in 'Dispose(bool disposing)' method
+        //     Dispose(disposing: false);
+        // }
+
+        public void Dispose()
+        {
+            // Do not change this code. Put cleanup code in 'Dispose(bool disposing)' method
+            Dispose(disposing: true);
+            GC.SuppressFinalize(this);
+        }
+
+        #endregion
+
+        #region Methods
+
+        public bool? GetState(ushort nodeNumber, ushort eventNumber)
```

Comment:
All the methods that are returning nullable things could instead use the `TryGetXX` pattern to avoid that situation - helps callers too because they don't need to compare against null etc

Reply:
Null in this context means that the state isn't known. One of those things that I'd be putting in the comments; but hadn't got to.

Reply:
Yeah - just a styling choice around what's nicer to use as a consumer, ie:

```
var state = em.GetState(1,2);
if (state != null) {
  if (state.Value) {
    //do something with on
  } else {
    //do something with off
  }
}
```

vs 

```
if (em.TryGetState(1,2, out var state)) {
  if (state) {
    //do something with on
  } else {
    //do something with off
  }
}
```

Reply:
The best solution is of course to provide both forms. Far better to allow the library user to use which ever is more convenient and understandable for them that us trying to prescribe one over the other. Of course the other approach is to define an enum with three values: but that means that extra types need to be defined by us and used by the library user, and it still may not be a usable for them as either of the above approaches.